File: src/main/java/me/santio/utils/ChatUtils.java
package me.santio.utils;

import net.md_5.bungee.api.ChatColor;
import org.apache.commons.lang.WordUtils;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

@SuppressWarnings("unused")
public final class ChatUtils {
    private static final Pattern hexPattern = Pattern.compile("&#[a-fA-F0-9]{6}");
    
    public static String tacc(String text) {
        if(VersionUtils.getMajor() >= 16) {
            Matcher match = hexPattern.matcher(text);
            while (match.find()) {
                String color = text.substring(match.start(), match.end());
                text = text.replace(color, ChatColor.of(color.substring(1)).toString());
                match = hexPattern.matcher(text);
            }
        }
        return ChatColor.translateAlternateColorCodes('&', text);
    }
    
    public static String strip(String text) {
        return ChatColor.stripColor(text);
    }
    
    public static String toID(String text) {
        return strip(tacc(text)).replaceAll(" ", "_").toLowerCase();
    }
    
    public static String[] loreWrap(String description) {
        return WordUtils.wrap(description, 40).split("\n");
    }
}

File: src/main/java/me/santio/utils/text/ChatUtils.kt
@file:JvmName("ChatUtils")
package me.santio.utils.text

import net.md_5.bungee.api.ChatColor
import net.md_5.bungee.api.chat.BaseComponent
import net.md_5.bungee.api.chat.ClickEvent
import net.md_5.bungee.api.chat.HoverEvent
import net.md_5.bungee.api.chat.TextComponent
import net.md_5.bungee.api.chat.hover.content.Text
import org.bukkit.entity.Player
import java.util.*
import java.util.regex.Matcher

private val hexPattern = Regex("&(#[a-fA-F\\d]{6})").toPattern()

fun String.colored(): String {
    val matcher: Matcher = hexPattern.matcher(ChatColor.translateAlternateColorCodes('&', this))
    val buffer = StringBuffer()

    while (matcher.find()) matcher.appendReplacement(buffer, ChatColor.of(matcher.group(1)).toString())

    return matcher.appendTail(buffer).toString()
}

fun String.strip(): String {
    return ChatColor.stripColor(this)
}

fun String.generateID(): String {
    return this.colored().strip().replace(" ".toRegex(), "_").lowercase()
}

fun String.toComponent(): BaseComponent {
    return TextComponent(*TextComponent.fromLegacyText(this))
}

fun BaseComponent.hover(text: String): BaseComponent {
    this.hoverEvent = HoverEvent(HoverEvent.Action.SHOW_TEXT, Text(text))
    return this
}

fun BaseComponent.click(action: ClickEvent.Action, text: String): BaseComponent {
    this.clickEvent = ClickEvent(action, text)
    return this
}

fun BaseComponent.add(component: BaseComponent): BaseComponent {
    this.addExtra(component)
    return this
}

@Suppress("SpellCheckingInspection")
fun String.normalcase(): String {
    return this.split(" ", "-", "_").joinToString(" ") { s -> s.substring(0, 1).uppercase() + s.substring(1).lowercase() }
}

fun Player.sendMessage(message: Message) {
    this.spigot().sendMessage(message.component())
}